File: packages/shared/src/explore.ts
import type { PlaceSuggestion } from './japanSuggestions.js';
import { getDestinationProfile } from './destinations/profiles.js';

/** id'den türeyen kararlı puan (4.2–4.9). */
export function placeRating(place: PlaceSuggestion): number {
  if (place.rating != null) return place.rating;
  let h = 0;
  for (let i = 0; i < place.id.length; i++) h = (h * 31 + place.id.charCodeAt(i)) | 0;
  const frac = (Math.abs(h) % 8) / 10; // 0.0–0.7
  return Math.round((4.2 + frac) * 10) / 10;
}

/** Çocuk dostu: açık alan tanımlı değilse kategoriden türet. */
export function isKidFriendly(place: PlaceSuggestion): boolean {
  if (place.kidFriendly != null) return place.kidFriendly;
  return place.category === 'fun' || place.category === 'nature';
}

/** Google Haritalar yorum/arama linki (API key gerekmez). */
export function googleReviewsUrl(query: string): string {
  return `https://www.google.com/maps/search/?api=1&query=${encodeURIComponent(query)}`;
}

export interface FoodRecommendation {
  label: string;
  emoji?: string;
}

/** Anahtar kelimeye göre yemek emojisi seç. Eşleşme yoksa default 🍽️. */
function emojiForDish(name: string): string {
  const n = name.toLowerCase();
  if (n.includes('ramen')) return '🍜';
  if (n.includes('udon') || n.includes('soba')) return '🍲';
  if (n.includes('onigiri')) return '🍙';
  if (n.includes('sushi') || n.includes('sashimi') || n.includes('nigiri') || n.includes('omakase'))
    return '🍣';
  if (n.includes('tempura')) return '🍤';
  if (n.includes('takoyaki')) return '🐙';
  if (n.includes('okonomiyaki')) return '🥞';
  if (n.includes('yakitori')) return '🍢';
  if (n.includes('yakiniku') || n.includes('wagyu') || n.includes('beef'))
    return '🥩';
  if (n.includes('izakaya') || n.includes('sake') || n.includes('shochu')) return '🍶';
  if (n.includes('matcha') || n.includes('mochi') || n.includes('wagashi')) return '🍵';
  if (n.includes('curry')) return '🍛';
  if (n.includes('katsu')) return '🍱';
  if (n.includes('tonkatsu')) return '🍖';
  if (n.includes('kaiseki') || n.includes('teishoku')) return '🍱';
  if (n.includes('konbini') || n.includes('sokak') || n.includes('street')) return '🏪';
  if (n.includes('tonkotsu') || n.includes('shoyu') || n.includes('miso')) return '🍜';
  if (n.includes('dim sum') || n.includes('dumpling') || n.includes('gyoza')) return '🥟';
  if (n.includes('tatlı') || n.includes('dessert') || n.includes('crepe')) return '🍰';
  if (n.includes('kahve') || n.includes('coffee') || n.includes('latte')) return '☕';
  return '🍽️';
}

/** Ülke profilinden önerilen yemekler (cuisines + dishRecommendations). */
export function recommendedFoods(countryCode: string): FoodRecommendation[] {
  const profile = getDestinationProfile(countryCode);
  if (!profile) return [];
  const out: FoodRecommendation[] = [];
  const seen = new Set<string>();
  for (const c of profile.cuisines) {
    if (!seen.has(c.label)) {
      seen.add(c.label);
      out.push({ label: c.label, emoji: c.emoji ?? emojiForDish(c.label) });
    }
  }
  for (const dish of profile.dishRecommendations) {
    if (!seen.has(dish)) {
      seen.add(dish);
      out.push({ label: dish, emoji: emojiForDish(dish) });
    }
  }
  return out;
}

export function ratingStars(rating: number): string {
  const full = Math.floor(rating);
  const half = rating - full >= 0.5;
  return '★'.repeat(full) + (half ? '½' : '');
}

File: mobile/lib/domain/explore.dart
// TypeScript packages/shared/src/explore.ts'in Dart karşılığı.
// Keşif ekranı puanlama/etiket yardımcıları.

import 'destination_profiles.dart';
import 'japan_suggestions.dart';

/// id'den türeyen kararlı puan (4.2–4.9).
/// JS'teki `(h * 31 + code) | 0` 32-bit signed taşma davranışı birebir korunur.
double placeRating(PlaceSuggestion place) {
  if (place.rating != null) return place.rating!;
  var h = 0;
  for (final code in place.id.codeUnits) {
    h = (h * 31 + code) & 0xFFFFFFFF;
  }
  // 32-bit signed'a çevir (JS `| 0`).
  if (h >= 0x80000000) h -= 0x100000000;
  final frac = (h.abs() % 8) / 10; // 0.0–0.7
  return ((4.2 + frac) * 10).roundToDouble() / 10;
}

/// Çocuk dostu: açık alan tanımlı değilse kategoriden türet.
bool isKidFriendly(PlaceSuggestion place) {
  if (place.kidFriendly != null) return place.kidFriendly!;
  return place.category == 'fun' || place.category == 'nature';
}

/// Google Haritalar yorum/arama linki (API key gerekmez).
String googleReviewsUrl(String query) =>
    'https://www.google.com/maps/search/?api=1&query=${Uri.encodeComponent(query)}';

class FoodRecommendation {
  const FoodRecommendation({required this.label, this.emoji});
  final String label;
  final String? emoji;
}

/// Anahtar kelimeye göre yemek emojisi seç. Eşleşme yoksa default 🍽️.
String _emojiForDish(String name) {
  final n = name.toLowerCase();
  if (n.contains('ramen')) return '🍜';
  if (n.contains('udon') || n.contains('soba')) return '🍲';
  if (n.contains('onigiri')) return '🍙';
  if (n.contains('sushi') ||
      n.contains('sashimi') ||
      n.contains('nigiri') ||
      n.contains('omakase')) {
    return '🍣';
  }
  if (n.contains('tempura')) return '🍤';
  if (n.contains('takoyaki')) return '🐙';
  if (n.contains('okonomiyaki')) return '🥞';
  if (n.contains('yakitori')) return '🍢';
  if (n.contains('yakiniku') || n.contains('wagyu') || n.contains('beef')) {
    return '🥩';
  }
  if (n.contains('izakaya') || n.contains('sake') || n.contains('shochu')) {
    return '🍶';
  }
  if (n.contains('matcha') || n.contains('mochi') || n.contains('wagashi')) {
    return '🍵';
  }
  if (n.contains('curry')) return '🍛';
  if (n.contains('katsu')) return '🍱';
  if (n.contains('tonkatsu')) return '🍖';
  if (n.contains('kaiseki') || n.contains('teishoku')) return '🍱';
  if (n.contains('konbini') || n.contains('sokak') || n.contains('street')) {
    return '🏪';
  }
  if (n.contains('tonkotsu') || n.contains('shoyu') || n.contains('miso')) {
    return '🍜';
  }
  if (n.contains('dim sum') || n.contains('dumpling') || n.contains('gyoza')) {
    return '🥟';
  }
  if (n.contains('tatlı') || n.contains('dessert') || n.contains('crepe')) {
    return '🍰';
  }
  if (n.contains('kahve') || n.contains('coffee') || n.contains('latte')) {
    return '☕';
  }
  return '🍽️';
}

/// Ülke profilinden önerilen yemekler (cuisines + dishRecommendations).
List<FoodRecommendation> recommendedFoods(String countryCode) {
  final profile = getDestinationProfile(countryCode);
  if (profile == null) return [];
  final out = <FoodRecommendation>[];
  final seen = <String>{};
  for (final c in profile.cuisines) {
    if (seen.add(c.label)) {
      out.add(FoodRecommendation(label: c.label, emoji: c.emoji));
    }
  }
  for (final dish in profile.dishRecommendations) {
    if (seen.add(dish)) {
      out.add(FoodRecommendation(label: dish, emoji: _emojiForDish(dish)));
    }
  }
  return out;
}

String ratingStars(double rating) {
  final full = rating.floor();
  final half = rating - full >= 0.5;
  return '★' * full + (half ? '½' : '');
}
